decisions/DR013-Project-structure.md

```diff
@@ -0,0 +1,350 @@
+  `src/python/PyImath*Test`.
+  - Python tests are old-school `assert`-based pure Python files (but
+    with extension `.in`) and executed as CTest tests.
+* Docs (Doxygen, Sphinx) under `docs`
+* CI/build helpers under repo-root `share/ci/scripts`, `cmake` and
+  `config` directories.
+
+#### [OpenColorIO](https://github.com/AcademySoftwareFoundation/OpenColorIO)
+
+* C++/Python/Java project. Multiple components.
+* `setup.py` at repo root that drives `CMakeLists.txt` at repo root.
+* Code grouped by component under `src` directory.
+  - Sub-groupings under subdirectories too, e.g. `apps`, `bindings`.
+  - Public headers under repo-root `include` directory.
+* No pure Python code. Python bindings located under
+  `src/bindings/python`. Similarly for Java.
+* Tests located under repo-root `tests`.
+  - Subdirectories split core tests, test utils and tests of optional
+    components, with `cpu` containing the core C++ tests and `python`
+    containing pure Python tests.
+* Docs (Doxygen, Sphinx) under `docs`.
+  - ASWF meeting notes under `ASWF`.
+* CI/build/docs helper scripts under subdirectories of repo-root `share`
+  directory.
+* Plugins for DCC tools under subdirectories of the repo-root `vendor`
+  directory, with additional scripts under subdirectories of the
+  repo-root `share` directory.
+
+## Options considered
+
+### Option 1
+
+Current structure.
+
+```
+.
+├── cmake
+├── contributing
+├── decisions
+├── doc
+│   └── src
+├── logo
+├── python
+│   └── *.py
+├── resources
+│   ├── build
+│   ├── examples
+│   │   ├── host
+│   │   │   └── simpleResolver
+│   │   └── manager
+│   │       └── SampleAssetManager
+│   └── openassetio-traitgen
+├── src
+│   ├── openassetio-core
+│   │   ├── include
+│   │   │   └── *.hpp
+│   │   └── *.cpp
+│   ├── openassetio-core-c
+│   │   ├── include
+│   │   │   └── *.h
+│   │   └── *.c
+│   └── openassetio-python
+|       ├── bridge
+│       │   ├── include
+│       │   │   └── *.hpp
+│       │   ├── *.cpp
+│       ├── module
+│       │   └── *.cpp
+│       └── private
+│           └── include
+│               └── *.hpp
+├── tests
+│   ├── openassetio-core
+│   │   └── *.cpp
+│   ├── openassetio-core-c
+│   │   └── *.cpp
+│   ├── openassetio-python
+│   │   └── *.cpp
+│   └── python
+│       └── *.py
+└── thirdparty
+    └── tomlplusplus
+```
+
+#### Pros
+
+ - No work to do.
+
+#### Cons
+
+- Public header paths (`include`) are mixed up with private source
+  paths, despite them being semantically distinct.
+- Bundling some third-party dependencies and supplying others externally
+  is inconsistent. This was only done to support building on an
+  unmodified ASWF Docker image, which is not a compelling reason.
+- Duplication of `if (OPENASSETIO_ENABLE_<COMPONENT>)` under `tests`
+
+Estimated cost: Small
+
+
+### Option 2
+
+* Place pure Python component under `src/openassetio-python/python`
+  directory, excluding `setup.py` and `pyproject.toml`.
```

Comment:
[question] This used to say "**including** `setup.py`", but I changed this based on recent learnings.  As mentioned in the fixup commit (3a79b6fd)

> `pip install .` won't work out of the box otherwise, since the directory containing `setup.py` is copied for the build process, breaking any relative paths.

Reply:
It still seems desirable to have `setup.py` able to sit inside the python subdirectory, for all the same reasons of project separation that we discussed.

I'm not sure we need to go to any effort to do it right now though, as further project seperation may never occur. We could log an issue for it as potential future work.

Reply:
Added #728, how's that?